extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryFrameProcessor.java

```diff
@@ -152,6 +164,62 @@ public void cleanup() throws IOException
     closer.close();
   }
 
+  public static Sequence<Object[]> mappingFunction(Sequence<ScanResultValue> inputSequence)
+  {
+    return inputSequence.flatMap(resultRow -> {
+      List<List<Object>> events = (List<List<Object>>) resultRow.getEvents();
+      return Sequences.simple(events);
+    }).map(List::toArray);
+  }
+
+  @Override
+  protected ReturnOrAwait<Unit> runWithLoadedSegment(final SegmentWithDescriptor segment) throws IOException
+  {
+    if (cursor == null) {
+      final Pair<LoadedSegmentDataProvider.DataServerQueryStatus, Yielder<Object[]>> statusSequencePair =
+          segment.fetchRowsFromDataServer(
+              query,
+              ScanQueryFrameProcessor::mappingFunction,
+              closer
+          );
+      if (LoadedSegmentDataProvider.DataServerQueryStatus.HANDOFF.equals(statusSequencePair.lhs)) {
+        log.info("Segment[%s] was handed off, falling back to fetching the segment from deep storage.",
+                 segment.getDescriptor());
+        return runWithSegment(segment);
+      }
+
+      RowSignature rowSignature = ScanQueryKit.getAndValidateSignature(query, jsonMapper);
+      Pair<Cursor, Closeable> cursorFromIterable = IterableRowsCursorHelper.getCursorFromYielder(
```

Comment:
Please close the closeable from cursorFromInterable. 

Reply:
Done! Missed this while resolving merge conflicts